design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+
+## Discussion
+
+### Chunks url styles
+
+* Many chunks from a few objects, the prefix can be the file
+  * Example: `s3://some-bucket/some-prefix/some-file.nc` is pointed to by 10k chunks.
+  * Having between 10 and 10k chunks from the same file is not unusual
+* Virtual zarr: each chunk comes from a different object, the prefix is not the object, we still need an extra string
+  * Example: `s3://some-bucket/some-prefix/c/0/0/1` is pointed to by only 1 chunk. But there are many objects like it.
+  * There could be tens of millions of chunks (so different urls)
+
+### Prefixes
+
+Storing the full URL for each chunk duplicates a lot of data in the manifest, making it large. The idea is to try to extract common _prefixes_ and identify them with a short id. Then the manifest only needs to identify the prefix id + maybe an extra string + length + offset.
+
+In the first url style above, the prefix could be the full file, `s3://some-bucket/some-prefix/some-file.nc`. In the second example, there are options for the prefix, it could be `s3://some-bucket/some-prefix/c/0/0/1`, or `s3://some-bucket/some-prefix/c/0/0` or `s3://some-bucket/some-prefix/c/0` or `s3://some-bucket/some-prefix/c`, with different numbers of chunks under it in each case.
+
+Intelligence is needed to optimize what is consider a good prefix. If done badly, we could have:
+
+* as many prefixes as chunks, for example by selecting `s3://some-bucket/some-prefix/c/0/0/1` in the second example above
+* too large of an extra string in each chunk, for example, if we select `s3://some-bucket` as the prefix each chunk needs to add something like `some-prefix/c/0/0`.
+
+### Should prefixes be per snapshot or global?
+
+The advantage of per snapshot is that we don't need to load/parse all the prefixes if the current snapshot uses only a few. A case for this would be a virtual dataset that has been materialized in a newer version. If prefixes are per-snapshot we wouldn't need to load any prefixes.
+
+On the other hand, requirement 4 asks for the full list of "dependencies" for the dataset, and that means for all versions, which requires a global list of prefixes. Without a global list, we would need to fetch the list of prefixes of each snapshot.
+
+### Who creates/decides the prefixes
+
+Obviously, generating the list of prefixes given the list of virtual chunks is doable automatically. But it's a tricky optimization problem. One strategy could be to generate a trie with all chunk urls, and then use some heuristic to select prefixes using heavily populated nodes. It's tricky and potentially computationally expensive for large repos.
+
+An alternative is to let the user dictate the prefixes. This way, they can optimize them knowing beforehand what virtual chunks they will add. If they don't optimize, they get large manifests without prefix extraction. We still need to satisfy requirements 2 and 4, which would require automatic extraction of at least the bucket and platform.
```

Comment:
I think it needs to be automatic.

Reply:
In principle it cannot be fully automated, it's not clear from a url in what platform, region, endpoint, etc a chunk lives. But I think the latest commit makes a much better job at making this easier. Configuration is optional and not needed in most normal cases.